### PR TITLE
Add Helm option to disable registering CRD from Cilium Operator

### LINKED ISSUE #15262: Add helm option to disable registering CRD from Cilium Operator
Cilium Operator is responsible for the creation of Cilium CRDs. We should have a helm option that allows users to disable this functionality.

### PR DESCRIPTION
**Pull request by Fedosin**
This PR contains three commits that: 
* add a Helm option `skipCRDCreation` for Cilium Operator that allows to skip CRDs creation;
* add `--skip-crd-creation` flag for the operator, indicating that CRDs creation should be skipped;
* deprecate `option.Config.SkipCRDCreation` and `--skip-crd-creation` for cilium-agent.

Fixes: #15262

### CODE REVIEW

File: operator/main.go [hunk 1]
```diff
@@ -274,8 +274,12 @@ func runOperator() {
...
 	// Register the CRDs after validating that we are running on a supported
 	// version of K8s.
-	if err := k8s.RegisterCRDs(); err != nil {
-		log.WithError(err).Fatal("Unable to register CRDs")
+	if !operatorOption.Config.SkipCRDCreation {
+		if err := k8s.RegisterCRDs(); err != nil {
+			log.WithError(err).Fatal("Unable to register CRDs")
+		}
+	} else {
+		log.Info("Skipping creation of CRDs")
```
Comment (aanm) [inline operator/main.go]: Hmm, maybe it's a good idea to move `operatorOption.Config.SkipCRDCreation` to inside the `RegisterCRDs` function and remove the other option is in there (`option.Config.SkipCRDCreation`). As well as remove the `SkipCRDCreation` variable from the option.Config altogether.

**Review by aanm — CHANGES_REQUESTED** [review]

File: operator/flags.go [hunk 2]
```diff
@@ -315,5 +315,8 @@ func init() {
 	flags.String(option.BGPConfigPath, "/var/lib/cilium/bgp/config.yaml", "Path to file containing the BGP configuration")
 	option.BindEnv(option.BGPConfigPath)
 
+	flags.Bool(option.SkipCRDCreation, false, "Skip Kubernetes Custom Resource Definitions creations")
```
Comment (kaworu) [inline operator/flags.go]: nit: I would reword the help message in a way that differ from what can be guessed by the flag name itself:
```suggestion
	flags.Bool(option.SkipCRDCreation, false, "when true, Kubernetes Custom Resource Definitions will not be created")
```

**Review by kaworu — CHANGES_REQUESTED** [review]
Hi @Fedosin and thank you for the PR! [CI failure](https://github.com/cilium/cilium/pull/15655/checks?check_run_id=2324388109) is legit, please run `make -C install/kubernetes docs`.

File: operator/main.go (hunk 1 above)
Comment (kaworu) [reply operator/main.go]: @aanm I wasn't able to parse your sentence, could you rephrase please?

**Review by kaworu — APPROVED** [review]
Helm related changes LGTM!

File: operator/main.go (hunk 1 above)
Comment (aanm) [reply operator/main.go]: @kaworu I've edited the comment.

Comment (Fedosin) [reply operator/main.go]: @aanm tbh, I thought about it, but `RegisterCRDs` is also [used by](https://docs.cilium.io/en/v1.9/cmdref/cilium-agent/) cilium-agent which also provides `--skip-crd-creation`. Since I'm not an expert in the code, I decided to keep the common function untouched.
If you think it is reasonable to do so, I can implement it in another PR.

Comment (aanm) [reply operator/main.go]: @Fedosin The `RegisterCRDs` are not used by the cilium-agents anymore, that's why I suggested to also remove the flag option from it.

Comment (Fedosin) [reply operator/main.go]: @aanm alright, if you think it is safe, then I'll update my PR later today

File: operator/flags.go [hunk 3]
```diff
@@ -315,5 +315,8 @@ func init() {
 	flags.String(option.BGPConfigPath, "/var/lib/cilium/bgp/config.yaml", "Path to file containing the BGP configuration")
 	option.BindEnv(option.BGPConfigPath)
 
+	flags.Bool(option.SkipCRDCreation, false, "when true, Kubernetes Custom Resource Definitions will not be created")
```
Comment (christarazi) [inline operator/flags.go]: This description should match the previous flag

File: operator/option/config.go [hunk 4]
```diff
@@ -41,6 +41,10 @@ const (
 	// compatible with MetalLB's configuration.
 	BGPConfigPath = "bgp-config-path"
 
+	// SkipCRDCreation specifies whether the CustomResourceDefinition will be
+	// disabled for the daemon
```
Comment (christarazi) [inline operator/option/config.go]: It's not for the daemon which refers to the cilium-agent though, so this should say Operator?

**Review by christarazi — CHANGES_REQUESTED** [review]

File: operator/flags.go (hunk 3 above)
Comment (Fedosin) [reply operator/flags.go]: @kaworu hi! should I revert your suggestion in this case? https://github.com/cilium/cilium/pull/15655#discussion_r611664659

File: operator/option/config.go (hunk 4 above)
Comment (Fedosin) [reply operator/option/config.go]: agree, updated

File: operator/flags.go (hunk 3 above)
Comment (christarazi) [reply operator/flags.go]: Ah, I didn't see that Alex has already suggested this. Then I think keeping it is fine.

```suggestion
	flags.Bool(option.SkipCRDCreation, false, "When true, Kubernetes Custom Resource Definitions will not be created")
```

Comment (Fedosin) [reply operator/flags.go]: done!

**Review by christarazi — APPROVED** [review]

File: daemon/cmd/daemon_main.go [hunk 5]
```diff
@@ -880,6 +880,7 @@ func init() {
 	option.BindEnv(option.SelectiveRegeneration)
 
 	flags.Bool(option.SkipCRDCreation, false, "Skip Kubernetes Custom Resource Definitions creations")
+	flags.MarkDeprecated(option.SkipCRDCreation, "This flag is no longer available and will be removed in v1.12")
```
Comment (aanm) [inline daemon/cmd/daemon_main.go]: This will be deprecated in 1.10, and the flag will be removed in 1.11.

**Review by aanm — CHANGES_REQUESTED** [review]
Also worth pointing out in the upgrade guide here https://github.com/cilium/cilium/blob/051032ce4e93121b94bc36cf1dfc8a51df0cc364/Documentation/operations/upgrade.rst#L401-L411

**Comment by aanm** [comment]
@Fedosin any update?

**Comment by errordeveloper** [comment]
hey @Fedosin do you think you'll get back to this sometime this week?

**Review by errordeveloper — CHANGES_REQUESTED** [review]
This PR now needs rebase, as @aanm pointed out.

**Comment by christarazi** [comment]
@Fedosin Please update the release note to have a summary sentence. No need to have a release note for each commit. The PR title is actually fine, so you can remove the entire release note section, if you'd like.

**Review by qmonnet — APPROVED** [review]

**Review by aanm — APPROVED** [review]

**Comment by aanm** [comment]
@Fedosin some documentation changes are required, please run `make -C Documentation update-cmdref` and add them to the commit. Thank you

**Review by christarazi — DISMISSED** [review]
Please squash your last commit into the commit that caused the change. No need for a separate commit

**Comment by errordeveloper** [comment]
@Fedosin I can see you updated the PR, did you squash per @christarazi's request?

**Comment by errordeveloper** [comment]
@Fedosin ping!

**Comment by Fedosin** [comment]
@errordeveloper yes, the commit has been squashed.

**Review by errordeveloper — APPROVED** [review]

**Comment by errordeveloper** [comment]
@Fedosin thanks  for confirming, looks like we are good to go once additional tests pass.
